cpp/src/arrow/testing/matchers.h

```diff
@@ -430,14 +430,10 @@ DataEqMatcher DataEqArray(T type, const std::vector<std::optional<ValueType>>& v
   static const bool need_safe_append = !is_fixed_width(T::type_id);
 
   for (auto value : values) {
-    if (value) {
-      if (need_safe_append) {
-        builder.UnsafeAppend(*value);
-      } else {
-        DCHECK_OK(builder.Append(*value));
-      }
+    if (need_safe_append) {
+      builder.UnsafeAppendOrNull(value);
```

Comment:
It's interesting that the `need_safe_append` use seems flipped. It probably should be:
```c++
    if (need_safe_append) {
      DCHECK_OK(builder.AppendOrNull(value));
    } else {
      builder.UnsafeAppendOrNull(value);
    }
```


@bkietz What do you think?

Reply:
I noticed that too. And decided to preserve the confusion. It would be great to fix it though by removing the `!` and then flipping the cases.

Reply:
(but that is assuming the current code is not buggy)

Reply:
I think it is.

Reply:
Swapped.